### PR TITLE
combining witness problem with ground checker

### PR DESCRIPTION
**Pull request by qcorporation**
adding test cases
removing visitNegation call within AstGroundAnalysis

### CODE REVIEW

File: src/AstSemanticChecker.cpp [hunk 1]
```diff
@@ -1015,15 +1013,13 @@ void AstSemanticCheckerImpl::checkIO() {
...
+    // Node-mapper that replaces aggregators with new (unique) variables and negations with boolean
+    // constraints of true
     struct M : public AstNodeMapper {
-        // Variables introduced to replace aggregators
-        mutable std::set<std::string> aggregatorVariables;
-
-        const std::set<std::string>& getAggregatorVariables() {
-            return aggregatorVariables;
+        mutable std::set<std::string> variables;
+        const std::set<std::string>& getVariables() {
```
Comment (ohamel-softwaresecure) [inline src/AstSemanticChecker.cpp]: why bother? `variables` is public.

File: src/AstSemanticChecker.cpp [hunk 2]
```diff
@@ -1032,26 +1028,27 @@ static const std::vector<SrcLocation> usesInvalidWitness(AstTranslationUnit& tu,
                 newVariableName << "+aggr_var_" << numReplaced++;
 
                 // Keep track of which variables are bound to aggregators
-                aggregatorVariables.insert(newVariableName.str());
+                variables.insert(newVariableName.str());
 
                 return std::make_unique<AstVariable>(newVariableName.str());
+            } else if (dynamic_cast<AstNegation*>(node.get())) {
+                // replace all negation terms with a boolean constraint of 'true'
```
Comment (ohamel-softwaresecure) [inline src/AstSemanticChecker.cpp]: Comment should explain *why* (and maybe *how*, if tricky) not *what*.

File: src/AstSemanticChecker.cpp [hunk 3]
```diff
@@ -1032,26 +1028,27 @@ static const std::vector<SrcLocation> usesInvalidWitness(AstTranslationUnit& tu,
...
     // Create two versions of the original clause
-
     // Clause 1 - will remain equivalent to the original clause in terms of variable groundedness
     auto originalClause = std::make_unique<AstClause>();
     originalClause->setHead(std::make_unique<AstAtom>("*"));
 
     // Clause 2 - will have aggregators replaced with intrinsically grounded variables
-    auto aggregatorlessClause = std::make_unique<AstClause>();
-    aggregatorlessClause->setHead(std::make_unique<AstAtom>("*"));
+    auto lessClause = std::make_unique<AstClause>();
```
Comment (ohamel-softwaresecure) [inline src/AstSemanticChecker.cpp]: Rename `erasedClause`?

File: src/AstSemanticChecker.cpp [hunk 4]
```diff
@@ -1524,6 +1496,26 @@ void GroundedTermsChecker::verify(AstTranslationUnit& translationUnit) {
...
+            bodyLiterals.push_back(groundnessConstraint.get());
+        });
+
+        // Perform the check
+        std::set<std::unique_ptr<AstArgument>> groundedArguments;
+        for (auto&& invalidArgument : usesInvalidWitness(translationUnit, bodyLiterals, groundedArguments)) {
+            report.addError(
+                    "Witness problem: argument grounded by an aggregator/negation's inner scope is used "
+                    "ungrounded in outer scope",
+                    invalidArgument);
```
Comment (ohamel-softwaresecure) [inline src/AstSemanticChecker.cpp]: grammar

File: tests/evaluation/neg8/neg8.dl [hunk 5]
```diff
@@ -0,0 +1,17 @@
+.decl b(x: number)
```
Comment (ohamel-softwaresecure) [inline tests/evaluation/neg8/neg8.dl:9]: Add "souffle devs 2020" header and a very short blurb explaining what behaviour this test case exercises.

**Review by ohamel-softwaresecure — CHANGES_REQUESTED** [review]
generally looks good. I'd've loved if the duplication work in check-witness could have been reduced, but that's not critical.

File: src/AstSemanticChecker.cpp (hunk 1 above)
Comment (qcorporation) [reply src/AstSemanticChecker.cpp]: remove getter

File: src/AstSemanticChecker.cpp (hunk 3 above)
Comment (qcorporation) [reply src/AstSemanticChecker.cpp]: renamed to `erasedClause`

File: tests/evaluation/neg8/neg8.dl (hunk 5 above)
Comment (qcorporation) [reply tests/evaluation/neg8/neg8.dl:9]: added header

File: src/AstSemanticChecker.cpp (hunk 4 above)
Comment (qcorporation) [reply src/AstSemanticChecker.cpp]: updated text

File: src/AstSemanticChecker.cpp (hunk 2 above)
Comment (qcorporation) [reply src/AstSemanticChecker.cpp]: updated comments

**Comment by ohamel-softwaresecure** [comment]
LGTM

**Review by ohamel-softwaresecure — APPROVED** [review]

**Pull request merged**